### PR TITLE
Fix for volumetric clouds performing Read and Write on same texture [Unsupported on some platforms]

### PR DESCRIPTION
**Pull request by FrancescoC-unity**
Fix for https://fogbugz.unity3d.com/f/cases/1316069/

As the title mentioned, clouds where not supported on some platforms (Metal and I assume Intel GPUs) as they were performing Read/Write on the same resource which format is 11-11-10. 

This change will perform an extra copy of the color buffer to use in those cases and sample that instead. 

I did try the following alternatives: 
 - Write full res clouds in the upsample without combining and then blending separately. [Slower]
 - Try to rewrite the shader from compute to pixel [Stop soon enough given that there are _a lot_ of samples so, LDS will be a big benefit] 

This option turned out to be the cheapest. It adds on PS4-base [where it won't run anyway, just used to profile] around 0.15ms of cost. 

**What did I test** forced the option on and off on both PC (NVIDIA card) and PS4 and check that same result is there. 

**NOTE FOR QA**: I don't have a dev-ready Metal machine to test, so we need to test on metal.

### CODE REVIEW

**Review by iM0ve — COMMENTED** [review]
**TL:DR**
Its now working but there are still some issues. Not sure if I should approve and we address them later or we want to fix them within this PR

**Whats tested:**
* Checked that every Volumetric Clouds parameter works on Windows
* Checked that every Volumetric Clouds parameter works on Mac
* Built player on mac to see that the clouds are present

Concerns:
* Only on Scene view, only on mac. There is this weird ring under the clouds
https://user-images.githubusercontent.com/36502659/116417194-b6ea0f80-a843-11eb-8621-88461618fd08.mov


* On my mac the framerate is really slow on a build. Its way faster in the editor
Build
https://user-images.githubusercontent.com/36502659/116417697-32e45780-a844-11eb-94cb-f0f0a6e2b37a.mov

Editor

https://user-images.githubusercontent.com/36502659/116418057-776ff300-a844-11eb-84dc-ce2420adc090.mov

**Comment by FrancescoC-unity** [comment]
> **TL:DR**
> Its now working but there are still some issues. Not sure if I should approve and we address them later or we want to fix them within this PR
> 
> **Whats tested:**
> 
> * Checked that every Volumetric Clouds parameter works on Windows
> * Checked that every Volumetric Clouds parameter works on Mac
> * Built player on mac to see that the clouds are present
> 
> Concerns:
> 
> * Only on Scene view, only on mac. There is this weird ring under the clouds
>   https://user-images.githubusercontent.com/36502659/116417194-b6ea0f80-a843-11eb-8621-88461618fd08.mov
> * On my mac the framerate is really slow on a build. Its way faster in the editor
>   Build
>   https://user-images.githubusercontent.com/36502659/116417697-32e45780-a844-11eb-94cb-f0f0a6e2b37a.mov
> 
> Editor
> 
>  Screen.Recording.2021-04-28.at.17.08.53.mov

I'd say those are likely different issues I'd open new tickets for those.

**Comment by johnpars** [comment]
Hi! I just realized that this PR potentially fixes the same issue as this one: #4345 (however with the focus on MSAA). 
The solution in this PR is better (I perform hardware blending from an intermediate target into the MSAA target), and maybe we can expand it to handle the MSAA case and I can close the other one?

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/HDRenderPipeline.VolumetricClouds.cs [hunk 1]
```diff
@@ -444,13 +439,21 @@ VolumetricCloudsParameters PrepareVolumetricCloudsParameters(HDCamera hdCamera,
             parameters.historyValidity = historyValidity;
             parameters.planarReflection = (hdCamera.camera.cameraType == CameraType.Reflection);
 
+            parameters.needExtraColorBufferCopy = (GetColorBufferFormat() == GraphicsFormat.B10G11R11_UFloatPack32 &&
+                // On PC and Metal, but not on console.
+                (SystemInfo.graphicsDeviceType == GraphicsDeviceType.Direct3D11 ||
+                    SystemInfo.graphicsDeviceType == GraphicsDeviceType.Direct3D12 ||
+                    SystemInfo.graphicsDeviceType == GraphicsDeviceType.Metal ||
+                    SystemInfo.graphicsDeviceType == GraphicsDeviceType.Vulkan));
```
Comment (johnpars) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/HDRenderPipeline.VolumetricClouds.cs:447]: If we can add one more check here for MSAA, I think we can close #4345

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/HDRenderPipeline.VolumetricClouds.cs:447]: yes good point, MSAA fix should be integrate in this PR

**Comment by sebastienlagarde** [comment]
Note to self: redirect the PR to master once hd/bugfix is merge

**Review by sebastienlagarde — APPROVED** [review]

**Comment by FrancescoC-unity** [comment]
Actually @johnpars the problem is slightly different here as blending in place still is happening. I just avoid the Read+Write from same texture, but here the issue is still that with MSAA we'd need to change the UAV declaration to be multi-sampled and do per-sample blending. 

I think for MSAA your solution is probably still more efficient.

**Review by iM0ve — APPROVED** [review]
Today I opened my project again to file the fogbugz for concerns. The artefact is just gone and the framerate stabilized after I waited for ~20 seconds. So no more outstanding issues from my part. All looks good

**Comment by sebastienlagarde** [comment]
will wait for discussion with John
